core/domain/feature_flag_services_test.py

```diff
@@ -0,0 +1,417 @@
+            with self.swap_all_feature_names_set, self.swap(
+                feconf, 'ENV_IS_OPPIA_ORG_PRODUCTION_SERVER', True
+            ):
+                self.assertFalse(
+                    feature_services.is_feature_flag_enabled(
+                        self.owner_id, self.dev_feature.name))
+
+    def test_evaluate_test_feature_for_prod_server_returns_false(self) -> None:
+        with self.swap_all_feature_flags, self.swap(
+            constants, 'DEV_MODE', False
+        ):
+            with self.swap_all_feature_names_set, self.swap(
+                feconf, 'ENV_IS_OPPIA_ORG_PRODUCTION_SERVER', True
+            ):
+                self.assertFalse(
+                    feature_services.is_feature_flag_enabled(
+                        self.owner_id, self.test_feature.name))
+
+    def test_evaluate_prod_feature_for_prod_server_returns_true(self) -> None:
+        with self.swap_all_feature_flags, self.swap(
+            constants, 'DEV_MODE', False
+        ):
+            with self.swap_all_feature_names_set, self.swap(
+                feconf, 'ENV_IS_OPPIA_ORG_PRODUCTION_SERVER', True
+            ):
+                self.assertTrue(
+                    feature_services.is_feature_flag_enabled(
+                        self.owner_id, self.prod_feature.name))
+
+    def test_feature_flag_is_enabled_when_force_enable_is_set_to_true(
+        self) -> None:
+        self.assertTrue(feature_services.is_feature_flag_enabled(
+            self.owner_id, self.dev_feature.name))
+
+    def test_feature_is_enabled_for_logged_out_users_with_force_enable_prop(
+        self) -> None:
+        self.assertTrue(feature_services.is_feature_flag_enabled(
+            None, self.dev_feature.name))
+
+    def test_feature_not_enabled_for_loggedout_user(self) -> None:
+        with self.swap_all_feature_names_set:
+            feature_services.update_feature_flag(
+                self.dev_feature.name, False, 0, [])
+        self.assertFalse(feature_services.is_feature_flag_enabled(
+            None, self.dev_feature.name))
+
+    def _signup_multiple_users_and_return_ids(self) -> Tuple[str, str, str]:
+        """Signup multiple users and returns user ids of them
+
+        Returns:
+            Tuple[int, int, int]. The tuple of user ids.
+        """
+        user_1_email = 'user1@example.com'
+        user_1_username = 'username1'
+        self.signup(user_1_email, user_1_username)
+        user_1_id = self.get_user_id_from_email(user_1_email)
+        user_2_email = 'user2@example.com'
+        user_2_username = 'username2'
+        self.signup(user_2_email, user_2_username)
+        user_2_id = self.get_user_id_from_email(user_2_email)
+        user_3_email = 'user3@example.com'
+        user_3_username = 'username3'
+        self.signup(user_3_email, user_3_username)
+        user_3_id = self.get_user_id_from_email(user_3_email)
+
+        return (user_1_id, user_2_id, user_3_id)
+
+    def _get_feature_status_for_users(
```

Comment:
This isn't well-named. There isn't any indication that it relates to dev_feature specifically, so it looks more general than it actually is.

Why do we need so much logic in the test support functions anyway? Can't this logic go directly in the code for the individual test since it seems quite test-specific (it's not really generic util functionality)?

Reply:
I have updated this.